File: src/Ethereum/MessageSigner.cpp
// SPDX-License-Identifier: Apache-2.0
//
// Copyright © 2017 Trust Wallet.

#include "MessageSigner.h"
#include <proto/Ethereum.pb.h>
#include <proto/TransactionCompiler.pb.h>
#include "rust/Wrapper.h"
#include "TrustWalletCore/TWCoinType.h"

namespace TW::Ethereum {

std::string signMessageRust(const PrivateKey& privateKey, const std::string& message, Proto::MessageType msgType, MessageSigner::MaybeChainId chainId) {
    Proto::MessageSigningInput input;
    input.set_private_key(privateKey.bytes.data(), privateKey.bytes.size());
    input.set_message(message);
    input.set_message_type(msgType);

    if (chainId.has_value()) {
        input.mutable_chain_id()->set_chain_id(static_cast<uint64_t>(chainId.value()));
    }

    Rust::TWDataWrapper inputData(data(input.SerializeAsString()));
    Rust::TWDataWrapper outputPtr = Rust::tw_message_signer_sign(inputData.get(), TWCoinTypeEthereum);

    auto outputData = outputPtr.toDataOrDefault();
    if (outputData.empty()) {
        return {};
    }

    Proto::MessageSigningOutput output;
    output.ParseFromArray(outputData.data(), static_cast<int>(outputData.size()));

    if (output.error() != Common::Proto::SigningError::OK) {
        return {};
    }
    return output.signature();
}

Data messagePreImageHashRust(const std::string& message, Proto::MessageType msgType) {
    Proto::MessageSigningInput input;
    input.set_message(message);
    input.set_message_type(msgType);

    Rust::TWDataWrapper inputData(data(input.SerializeAsString()));
    Rust::TWDataWrapper outputPtr = Rust::tw_message_signer_pre_image_hashes(inputData.get(), TWCoinTypeEthereum);

    auto outputData = outputPtr.toDataOrDefault();
    if (outputData.empty()) {
        return {};
    }

    TxCompiler::Proto::PreSigningOutput output;
    output.ParseFromArray(outputData.data(), static_cast<int>(outputData.size()));

    if (output.error() != Common::Proto::SigningError::OK) {
        return {};
    }
    return data(output.data_hash());
}

bool verifyMessageRust(const PublicKey& publicKey, const std::string& message, const std::string& signature) {
    Proto::MessageVerifyingInput input;
    input.set_public_key(publicKey.bytes.data(), publicKey.bytes.size());
    input.set_message(message);
    input.set_signature(signature);

    Rust::TWDataWrapper inputData(data(input.SerializeAsString()));
    return Rust::tw_message_signer_verify(inputData.get(), TWCoinTypeEthereum);
}

std::string MessageSigner::signMessage(const PrivateKey& privateKey, const std::string& message, MessageType msgType, MaybeChainId chainId) {
    auto protoMsgType = Proto::MessageType::MessageType_legacy;
    switch (msgType) {
        case MessageType::Eip155: {
            protoMsgType = Proto::MessageType::MessageType_eip155;
            break;
        }
        case MessageType::ImmutableX: {
            protoMsgType = Proto::MessageType::MessageType_immutable_x;
            break;
        }
        default: {
            break;
        }
    }

    return signMessageRust(privateKey, message, protoMsgType, chainId);
}

std::string MessageSigner::signTypedData(const PrivateKey& privateKey, const std::string& data, MessageType msgType, MessageSigner::MaybeChainId chainId) {
    auto protoMsgType = Proto::MessageType::MessageType_typed;
    switch (msgType) {
        case MessageType::Eip155: {
            protoMsgType = Proto::MessageType::MessageType_typed_eip155;
            break;
        }
        default: {
            break;
        }
    }

    return signMessageRust(privateKey, data, protoMsgType, chainId);
}

bool MessageSigner::verifyMessage(const PublicKey& publicKey, const std::string& message, const std::string& signature) noexcept {
    return verifyMessageRust(publicKey, message, signature);
}

Data MessageSigner::messagePreImageHash(const std::string& message) noexcept {
    return messagePreImageHashRust(message, Proto::MessageType::MessageType_legacy);
}

Data MessageSigner::typedDataPreImageHash(const std::string& data) noexcept {
    return messagePreImageHashRust(data, Proto::MessageType::MessageType_typed);
}

void MessageSigner::prepareSignature(Data& signature, MessageType msgType, TW::Ethereum::MessageSigner::MaybeChainId chainId) noexcept {
    switch (msgType) {
        case MessageType::ImmutableX: {
            break;
        }
        case MessageType::Legacy: {
            signature[64] += 27;
            break;
        }
        case MessageType::Eip155: {
            auto id = chainId.value_or(0);
            signature[64] += 35 + id * 2;
            break;
        }
        default:
            break;
    }
}

} // namespace TW::Ethereum

File: rust/tw_evm/src/modules/message_signer.rs
// SPDX-License-Identifier: Apache-2.0
//
// Copyright © 2017 Trust Wallet.

use crate::message::eip191::Eip191Message;
use crate::message::eip712::eip712_message::Eip712Message;
use crate::message::signature::{MessageSignature, SignatureType};
use crate::message::{to_signing, EthMessage, EthMessageBoxed};
use std::borrow::Cow;
use std::str::FromStr;
use tw_coin_entry::coin_context::CoinContext;
use tw_coin_entry::error::prelude::*;
use tw_coin_entry::modules::message_signer::MessageSigner;
use tw_coin_entry::signing_output_error;
use tw_encoding::hex::ToHex;
use tw_keypair::ecdsa::secp256k1;
use tw_keypair::ecdsa::signature::VerifySignature;
use tw_keypair::traits::{SigningKeyTrait, VerifyingKeyTrait};
use tw_number::U256;
use tw_proto::Ethereum::Proto;
use tw_proto::TxCompiler::Proto as CompilerProto;

#[derive(Default)]
pub struct EthMessageSigner;

impl MessageSigner for EthMessageSigner {
    type MessageSigningInput<'a> = Proto::MessageSigningInput<'a>;
    type MessagePreSigningOutput = CompilerProto::PreSigningOutput<'static>;
    type MessageSigningOutput = Proto::MessageSigningOutput<'static>;
    type MessageVerifyingInput<'a> = Proto::MessageVerifyingInput<'a>;

    fn message_preimage_hashes(
        &self,
        _coin: &dyn CoinContext,
        input: Self::MessageSigningInput<'_>,
    ) -> Self::MessagePreSigningOutput {
        Self::message_preimage_hashes_impl(input)
            .unwrap_or_else(|e| signing_output_error!(CompilerProto::PreSigningOutput, e))
    }

    fn sign_message(
        &self,
        _coin: &dyn CoinContext,
        input: Self::MessageSigningInput<'_>,
    ) -> Self::MessageSigningOutput {
        Self::sign_message_impl(input)
            .unwrap_or_else(|e| signing_output_error!(Proto::MessageSigningOutput, e))
    }

    fn verify_message(
        &self,
        _coin: &dyn CoinContext,
        input: Self::MessageVerifyingInput<'_>,
    ) -> bool {
        Self::verify_message_impl(input).unwrap_or_default()
    }
}

impl EthMessageSigner {
    fn message_preimage_hashes_impl(
        input: Proto::MessageSigningInput<'_>,
    ) -> SigningResult<CompilerProto::PreSigningOutput<'static>> {
        let msg = Self::message_from_proto(input)?;
        let hash = msg.hash().map_err(to_signing)?.to_vec();
        Ok(CompilerProto::PreSigningOutput {
            data: Cow::Owned(hash.clone()),
            data_hash: Cow::Owned(hash),
            ..CompilerProto::PreSigningOutput::default()
        })
    }

    fn sign_message_impl(
        input: Proto::MessageSigningInput<'_>,
    ) -> SigningResult<Proto::MessageSigningOutput<'static>> {
        let private_key = secp256k1::PrivateKey::try_from(input.private_key.as_ref())?;
        let signature_type =
            Self::signature_type_from_proto(input.message_type, input.chain_id.clone());

        let msg = Self::message_from_proto(input)?;

        let hash_to_sign = msg.hash().map_err(to_signing)?;
        let secp_sign = private_key.sign(hash_to_sign)?;
        let prepared_sign = MessageSignature::prepared(secp_sign, signature_type)?;

        Ok(Proto::MessageSigningOutput {
            signature: Cow::Owned(prepared_sign.to_bytes().to_hex()),
            ..Proto::MessageSigningOutput::default()
        })
    }

    fn verify_message_impl(input: Proto::MessageVerifyingInput<'_>) -> SigningResult<bool> {
        let public_key = secp256k1::PublicKey::try_from(input.public_key.as_ref())?;

        let msg_hash = Self::message_from_str(&input.message)?
            .hash()
            .map_err(to_signing)?;
        let secp_signature =
            MessageSignature::from_str(&input.signature)?.to_secp256k1_signature()?;

        let actual_public_key = secp256k1::PublicKey::recover(secp_signature.clone(), msg_hash)?;
        let valid = actual_public_key == public_key
            && public_key.verify(VerifySignature::from(secp_signature), msg_hash);

        Ok(valid)
    }

    fn message_from_proto(input: Proto::MessageSigningInput<'_>) -> SigningResult<EthMessageBoxed> {
        match input.message_type {
            Proto::MessageType::MessageType_legacy
            | Proto::MessageType::MessageType_eip155
            | Proto::MessageType::MessageType_immutable_x => {
                Ok(Eip191Message::new(input.message).into_boxed())
            },
            Proto::MessageType::MessageType_typed
            | Proto::MessageType::MessageType_typed_eip155 => match input.chain_id {
                Some(expected_chain_id) => {
                    let expected_chain_id = U256::from(expected_chain_id.chain_id);
                    Ok(Eip712Message::new_checked(input.message, expected_chain_id)
                        .map_err(to_signing)?
                        .into_boxed())
                },
                None => Ok(Eip712Message::new(input.message)
                    .map_err(to_signing)?
                    .into_boxed()),
            },
        }
    }

    fn message_from_str(user_message: &str) -> SigningResult<EthMessageBoxed> {
        match Eip712Message::new(user_message) {
            Ok(typed_data) => Ok(typed_data.into_boxed()),
            Err(_) => Ok(Eip191Message::new(user_message).into_boxed()),
        }
    }

    fn signature_type_from_proto(
        msg_type: Proto::MessageType,
        maybe_chain_id: Option<Proto::MaybeChainId>,
    ) -> SignatureType {
        match msg_type {
            Proto::MessageType::MessageType_immutable_x => SignatureType::Standard,
            Proto::MessageType::MessageType_legacy | Proto::MessageType::MessageType_typed => {
                SignatureType::Legacy
            },
            Proto::MessageType::MessageType_eip155
            | Proto::MessageType::MessageType_typed_eip155 => {
                let chain_id = U256::from(maybe_chain_id.unwrap_or_default().chain_id);
                SignatureType::Eip155 { chain_id }
            },
        }
    }
}
